### DISCUSSION TITLE
detecting when DMA finished transfer of whole bytearray

### DISCUSSION THREAD

**Discussion by jkozniewski** [micropython/micropython#15114] [RP2040 / Pico]
Hello - I have managed to setup DMA and PIO in such way that DMA feeds PIO with data from bytearray 4 bytes at a time,
and I wish to get interupt when whole bytearray has beed "processed".
```python
d = rp2.DMA()
d.irq( lambda p: print("dma irq!") )

DATA_REQUEST_INDEX = (0 << 3) + 0 #state machine index 

# Transfer bytes, rather than words, don't increment the write address and pace the transfer.
c = d.pack_ctrl(size=2, inc_write=False, treq_sel=DATA_REQUEST_INDEX, irq_quiet=False) 

d.config(
    read=src_data, #my bytearray
    write=sm, #my statemachine
    count=len(src_data)//4, #transfer count is bytes / 4 since I'm pushing 32bits at a time to FIFO
    ctrl=c,
    trigger=True
)
```

The interrupt indeed gets fired but after the _first_ transfer not after the last one.
I know that since I also have interrupt raised from PIO code of my StateMachine which is raised on each new 32bits pushed into FIFO and the message from DMA irq is placed after first message from PIO irq. 
I'm also checking the `DMA.count` property and it's 1 after I read first 32bits and 0 for all subsequent transfers.

So - how can I verify if indeed the whole bytearray has been  pushed from DMA to PIO, am I getting the DMA IRQ and `count` property wrong ?

**Comment by GitHubsSilverBullet** [comment]
Don't know how large your source array is, but maybe it's pushed entirely into the SM-FIFO. That'd be blazingly fast.
Check you DMA transmit count whether it's already down to zero.

**Reply by jkozniewski** [reply]
Well - as I understand it should be pushed either 8, 16 or 32bits at the time paced by DREQ ( treq_sel ) of PIO, so I should observe the count being equal to `count` parameter in `config` getting down to 0 not being 1 and then immediately 0... ?
Maybe that's terminology thing - what is considered a _count of transfers_ - amount of "chunks" from the source or the whole source is considered a single transfer (but that wouldn't make much sense to me).
